File: frontend/lib/targets.ts
/**
 * Target registry types and surface helpers.
 *
 * Mirrors `backend/src/core/models/targets.py` — one source of truth for the
 * shape, kept in sync by hand as with the rest of `lib/types.ts`.
 */

export type TargetKind =
  | "llm"
  | "chatbot"
  | "rag"
  | "agent"
  | "multi_agent"
  | "mcp_agent"
  | "api_app"
  | "workflow";

export const TARGET_KINDS: ReadonlyArray<{ id: TargetKind; label: string }> = [
  { id: "llm", label: "LLM" },
  { id: "chatbot", label: "Chatbot" },
  { id: "rag", label: "RAG application" },
  { id: "agent", label: "AI agent" },
  { id: "multi_agent", label: "Multi-agent system" },
  { id: "mcp_agent", label: "MCP-connected agent" },
  { id: "api_app", label: "API application" },
  { id: "workflow", label: "Enterprise workflow" },
];

export type CapabilityId =
  | "tools"
  | "memory"
  | "rag"
  | "mcp"
  | "external_api"
  | "system_prompt_disclosure"
  | "input_guardrail"
  | "output_guardrail";

export const CAPABILITY_LABELS: Record<CapabilityId, string> = {
  tools: "Tools",
  memory: "Memory",
  rag: "Retrieval",
  mcp: "MCP / plugins",
  external_api: "Network access",
  system_prompt_disclosure: "Prompt disclosure",
  input_guardrail: "Input guardrail",
  output_guardrail: "Output guardrail",
};

export interface TargetCapability {
  id: CapabilityId;
  present: boolean;
  confidence: number;
  evidence: string;
}

export interface TargetSurfaceItem {
  taxonomy_id: string;
  title: string;
  rationale: string;
  from_capabilities: string[];
}

export interface TargetSurface {
  reachable: boolean;
  unreachable_reason: string | null;
  probes_run: number;
  reported_model: string | null;
  capabilities: TargetCapability[];
  surface: TargetSurfaceItem[];
  trust_boundaries: string[];
  discovered_at: string | null;
}

export interface Target {
  id: string;
  tenant_id: string;
  name: string;
  kind: TargetKind;
  version: string;
  description: string | null;
  provider: string;
  model: string;
  base_url: string | null;
  system_prompt: string | null;
  authorized: boolean;
  tags: string[];
  config: Record<string, unknown>;
  surface: TargetSurface | null;
  discovered_at: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface TargetDraft {
  name: string;
  kind: TargetKind;
  version: string;
  description?: string | null;
  provider: string;
  model: string;
  base_url?: string | null;
  system_prompt?: string | null;
  authorized: boolean;
}

export function kindLabel(kind: string): string {
  return TARGET_KINDS.find((k) => k.id === kind)?.label ?? kind;
}

/** Where a target sits in the discovery lifecycle. Drives the status chip. */
export type DiscoveryState = "unauthorized" | "undiscovered" | "unreachable" | "mapped";

export function discoveryState(target: Target): DiscoveryState {
  if (!target.authorized) return "unauthorized";
  if (!target.surface) return "undiscovered";
  if (!target.surface.reachable) return "unreachable";
  return "mapped";
}

export const DISCOVERY_STATE_COPY: Record<DiscoveryState, { label: string; hint: string }> = {
  unauthorized: {
    label: "Not authorized",
    hint: "Authorize this target before ARTSA sends it any traffic.",
  },
  undiscovered: {
    label: "Not mapped",
    hint: "Run discovery to find its capabilities and attack surface.",
  },
  unreachable: {
    label: "Unreachable",
    hint: "Discovery could not reach the target, so its surface is unknown.",
  },
  mapped: { label: "Mapped", hint: "Attack surface derived from discovery evidence." },
};

/** Capabilities discovery confirmed are present, in display order. */
export function presentCapabilities(surface: TargetSurface | null): TargetCapability[] {
  if (!surface) return [];
  const order = Object.keys(CAPABILITY_LABELS) as CapabilityId[];
  return surface.capabilities
    .filter((c) => c.present)
    .sort((a, b) => order.indexOf(a.id) - order.indexOf(b.id));
}

/**
 * Coverage of the 12-category taxonomy for one target.
 *
 * `open` counts categories the target's own capabilities exposed. An
 * unmapped target reports zero rather than a guess.
 */
export function surfaceCoverage(surface: TargetSurface | null): {
  open: number;
  total: number;
} {
  return { open: surface?.reachable ? surface.surface.length : 0, total: 12 };
}

File: backend/src/core/models/targets.py
"""Pydantic v2 models for the target registry and discovered attack surface."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

TargetKind = Literal[
    "llm",
    "chatbot",
    "rag",
    "agent",
    "multi_agent",
    "mcp_agent",
    "api_app",
    "workflow",
]

# Capabilities discovery looks for. Each one that is present widens the attack
# surface, so the surface report is derived from this set.
CapabilityId = Literal[
    "tools",
    "memory",
    "rag",
    "mcp",
    "external_api",
    "system_prompt_disclosure",
    "input_guardrail",
    "output_guardrail",
]


class Capability(BaseModel):
    """One discovered capability, with the evidence that led to the call."""

    id: CapabilityId
    present: bool
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    # Verbatim excerpt from the probe response that supports `present`.
    evidence: str = ""


class SurfaceItem(BaseModel):
    """An attack-surface entry implied by the discovered capabilities."""

    # ARTSA attack taxonomy id, e.g. "AI-01".
    taxonomy_id: str
    title: str
    rationale: str
    # Capability ids that opened this surface.
    from_capabilities: list[str] = Field(default_factory=list)


class TargetSurface(BaseModel):
    """Result of a discovery run against one target.

    ``reachable`` is false when the target could not be probed at all (no
    credentials, network failure). In that case capabilities stay empty rather
    than being guessed — an unreachable target has an *unknown* surface, not an
    empty one.
    """

    reachable: bool = False
    unreachable_reason: str | None = None
    probes_run: int = 0
    # Model identity as self-reported by the target, when it discloses one.
    reported_model: str | None = None
    capabilities: list[Capability] = Field(default_factory=list)
    surface: list[SurfaceItem] = Field(default_factory=list)
    trust_boundaries: list[str] = Field(default_factory=list)
    discovered_at: datetime | None = None


class TargetCreate(BaseModel):
    """Request body for registering a target."""

    name: str = Field(min_length=1, max_length=255)
    kind: TargetKind = "llm"
    version: str = Field(default="v1", max_length=64)
    description: str | None = Field(default=None, max_length=1024)
    provider: str = Field(default="", max_length=64)
    model: str = Field(default="", max_length=128)
    base_url: str | None = Field(default=None, max_length=1024)
    system_prompt: str | None = None
    authorized: bool = False
    tags: list[str] = Field(default_factory=list)
    config: dict[str, Any] = Field(default_factory=dict)


class TargetUpdate(BaseModel):
    """Partial update. Omitted fields are left untouched."""

    name: str | None = Field(default=None, min_length=1, max_length=255)
    kind: TargetKind | None = None
    version: str | None = Field(default=None, max_length=64)
    description: str | None = Field(default=None, max_length=1024)
    provider: str | None = Field(default=None, max_length=64)
    model: str | None = Field(default=None, max_length=128)
    base_url: str | None = Field(default=None, max_length=1024)
    system_prompt: str | None = None
    authorized: bool | None = None
    tags: list[str] | None = None
    config: dict[str, Any] | None = None


class Target(BaseModel):
    """A registered AI system under test."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    tenant_id: str
    name: str
    kind: str = "llm"
    version: str = "v1"
    description: str | None = None
    provider: str = ""
    model: str = ""
    base_url: str | None = None
    system_prompt: str | None = None
    authorized: bool = False
    tags: list[str] = Field(default_factory=list)
    config: dict[str, Any] = Field(default_factory=dict)
    surface: TargetSurface | None = None
    discovered_at: datetime | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None
